前处理
---
```
library(Seurat)
data <- Read10X(data.dir = './')
## 加入分组信息与否
### 不分组
sce = CreateSeuratObject(counts = data)
### 分组
if (F){
name1=data.frame(names=colnames(data),group=substr(colnames(data),18,18))
rownames(name1)=name1$names
name1$names=NULL
name1$group=gsub("1","control",name1$group)
name1$group=gsub("2","IS",name1$group)
name1$group=gsub("3","HS",name1$group)
## 载入seurat
sce = CreateSeuratObject(counts = data,meta.data = name1)
}

## 取想要的组
if(F){
sce=sce[,grepl("-2$",colnames(sce))]
}
## 获得ercc meta信息
ercc=rownames(sce)[grep("^ercc",rownames(sce),ignore.case = T)]
ercc=substr(ercc[1],0,4)
pbmc <- PercentageFeatureSet(sce, pattern = ercc, col.name = "percent.Ercc")
## 获得mt meta信息
mt=rownames(sce)[grep("^mt-",rownames(sce),ignore.case = T)]
mt=substr(mt[1],0,3)
pbmc <- PercentageFeatureSet(pbmc, pattern = mt, col.name = "percent.mt")
## 看占有率
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA","percent.Ercc",'percent.mt'), ncol = 4)
sce <- subset(pbmc, subset = nFeature_RNA <6000 & percent.Ercc < 0.3 & percent.mt<60)
## 得到细胞周期基因
library(homologene)
m.s.genes <- cc.genes.updated.2019$s.genes
m.g2m.genes <- cc.genes.updated.2019$g2m.genes
m.s.genes.mm=homologene(m.s.genes, inTax = 9606, outTax = 10090)
m.g2m.genes.mm=homologene(m.g2m.genes, inTax = 9606, outTax = 10090)
s.genes=m.s.genes.mm$`10090`
g2m.genes=m.g2m.genes.mm$`10090`
sce <- CellCycleScoring(sce, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
```
如果是多组单细胞，需要整合
---
- https://github.com/171909771/Singlecells/blob/main/seurat/%E6%95%B4%E5%90%88%E4%B8%8D%E5%90%8C%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE.md

如果只有一组单细胞，用下面的方法
---

方法1 SCtransform
---
```
pbmc <- SCTransform(sce, method = "glmGamPoi", vars.to.regress =  c("percent.Ercc","percent.mt","S.Score","G2M.Score"),conserve.memory = TRUE)
pbmc <- RunPCA(pbmc, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:30, verbose = FALSE)

pbmc <- FindNeighbors(pbmc, dims = 1:30, verbose = FALSE)
pbmc <- FindClusters(pbmc, verbose = FALSE)
DimPlot(pbmc, label = TRUE) + NoLegend()
```

方法2 归一化标准化
---
```
scenorm <- NormalizeData(sce, normalization.method = "LogNormalize", scale.factor = 10000)
scenormvc <- FindVariableFeatures(scenorm, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(scenormvc)
scaledata1 <- ScaleData(scenormvc, features = all.genes)
PCAdata1 <- RunPCA(scaledata1, features = VariableFeatures(object = scaledata1))
ddim1 <- JackStraw(PCAdata1, num.replicate = 100)
ddim2 <- ScoreJackStraw(ddim1, dims = 1:20)
JackStrawPlot(ddim2, dims = 1:20)
ElbowPlot(ddim2)
# 确定PCA 的维度
dim_n=14
cluster1 <- FindNeighbors(ddim2, dims = 1:dim_n)
cluster2 <- FindClusters(cluster1, resolution = seq(0.5,1.2,by=0.1))
library("clustree")
clustree(cluster2)
# 可以把这个参数加入到seurat数据中
Idents(object = cluster2) <- "RNA_snn_res.0.6"
umap1 <- RunUMAP(cluster2, dims = 1:dim_n)
DimPlot(umap1, reduction = "umap")
tsne1 <- RunTSNE(cluster2, dims = 1:dim_n)
DimPlot(tsne1, reduction = "tsne")
```
注释
---
```
library(SingleR)
library("celldex")
## 确定什么去注释
annotation=tsne1
sce_for_SingleR <- GetAssayData(annotation, slot="data")
clusters=annotation@active.ident
mouseRNA <- MouseRNAseqData()
pred.mouseRNA <- SingleR(test = sce_for_SingleR, ref = mouseRNA, labels = mouseRNA$label.fine,
                         clusters = clusters, de.method="wilcox",
                         assay.type.test = "logcounts", assay.type.ref = "logcounts")
cellType=data.frame(ClusterID=levels(clusters),mouseRNA=pred.mouseRNA$labels)
# 可视化看注释效果
plotScoreHeatmap(pred.mouseRNA, clusters=pred.mouseRNA@rownames, fontsize.row = 9,show_colnames = T)
new.cluster.ids <- pred.mouseRNA$labels
names(new.cluster.ids) <- levels(annotation)
annotation1 <- RenameIdents(annotation,new.cluster.ids)
# 改相应的分布名称TSNE OR UMAP
DimPlot(annotation1, reduction = "tsne", label = T)
```
